### PR TITLE
Fix cheats in library calls

### LINKED ISSUE #3880: Cheating with library call from test doesn't work
### Current State

When making a library call from the test code itself, cheats are not reflected in the called contract.

### Objective

Fix cheating for library calls

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by franciszekjob**
Closes #3880

## Introduced changes

Cheatcodes are now reflected in called contract, when directly using a library call

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Forge
+
+#### Fixed
+
+- Cheatcodes are now reflected in called contract, when directly using a library call
```
Comment (cptartur) [inline CHANGELOG.md]:
```suggestion
- Cheatcodes are now reflected in called contract, when directly using a library call from test code
```

File: crates/cheatnet/tests/cheatcodes/library_call.rs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+    assert_success(
+        test_env.library_call_contract(&class_hash, "get_caller_address", &[]),
+        &[Felt::from(cheated_caller_address)],
+    );
+
+    assert_success(
+        test_env.library_call_contract(&class_hash, "get_caller_address", &[]),
+        &[Felt::from(cheated_caller_address)],
+    );
```
Comment (cptartur) [inline crates/cheatnet/tests/cheatcodes/library_call.rs:31]: Why do we call twice here?

File: crates/cheatnet/tests/cheatcodes/library_call.rs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        CheatSpan::TargetCalls(1.try_into().unwrap()),
+    );
+
+    assert_success(
+        test_env.library_call_contract(&class_hash, "get_caller_address", &[]),
+        &[Felt::from(cheated_caller_address)],
+    );
+
+    // Library call doesn't trigger the cheat removal because it's not a contract call
+    // hence after second library call the cheated address is still in effect.
```
Comment (cptartur) [inline crates/cheatnet/tests/cheatcodes/library_call.rs:61]: I wonder if we every documented anywhere that library calls don't progress cheat spans (even though it makes sense)? If not, maybe we should add it in the docs somewhere?

File: crates/cheatnet/tests/cheatcodes/library_call.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    assert_success(
+        test_env.library_call_contract(&class_hash, "get_caller_address", &[]),
+        &[Felt::from(cheated_caller_address)],
+    );
+}
+
```
Comment (cptartur) [inline crates/cheatnet/tests/cheatcodes/library_call.rs:76]: Can we add a step where we stop the cheat here to check if it's reflected as well?

File: crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/execution_utils.rs [hunk 5]
```diff
@@ -16,11 +16,16 @@ pub(crate) fn resolve_cheated_data_for_call(
...
+        // Edge case, when delegate call is made by test contract
+        if cheatnet_state.trace_data.current_call_stack.size() == 1 {
+            cheatnet_state.get_cheated_data(entry_point.caller_address)
+        } else {
+            cheatnet_state
+                .trace_data
+                .current_call_stack
+                .top_cheated_data()
+                .clone()
+        }
```
Comment (cptartur) [inline crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/execution_utils.rs]: I think we could benefit for bit longer comment explaining what's the interaction here and why we need this.

File: crates/cheatnet/tests/cheatcodes/library_call.rs (hunk 2 above)
Comment (franciszekjob) [reply crates/cheatnet/tests/cheatcodes/library_call.rs:31]: Because I wanted to show that cheat is reflected after calling more than once, as it's global (so it's indefinite too).
Could be changed if you find it misleading or smth.

File: crates/cheatnet/tests/cheatcodes/library_call.rs (hunk 4 above)
Comment (franciszekjob) [reply crates/cheatnet/tests/cheatcodes/library_call.rs:76]: https://github.com/foundry-rs/starknet-foundry/pull/3958/commits/5c00b1ab0af99c82c2e6215fd896b2c82bc797a4

File: crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/execution_utils.rs (hunk 5 above)
Comment (franciszekjob) [reply crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/execution_utils.rs]: https://github.com/foundry-rs/starknet-foundry/pull/3958/commits/5ffa509a4c3abfc243ce713ae0b5a83f135a1c44

File: crates/cheatnet/tests/cheatcodes/library_call.rs (hunk 3 above)
Comment (franciszekjob) [reply crates/cheatnet/tests/cheatcodes/library_call.rs:61]: Sure, let's mention it. https://github.com/foundry-rs/starknet-foundry/pull/3958/commits/ec2b682d7cd6cbc62d612fc78fec3dbd42cce4dc

File: crates/cheatnet/tests/cheatcodes/library_call.rs (hunk 2 above)
Comment (cptartur) [reply crates/cheatnet/tests/cheatcodes/library_call.rs:31]: I'd just an comment if that's intentional, but it's okay to keep it.

File: crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/execution_utils.rs [hunk 6]
```diff
@@ -16,11 +16,18 @@ pub(crate) fn resolve_cheated_data_for_call(
     cheatnet_state: &mut CheatnetState,
 ) -> CheatedData {
     if let CallType::Delegate = entry_point.call_type {
-        cheatnet_state
-            .trace_data
-            .current_call_stack
-            .top_cheated_data()
-            .clone()
+        // This is an edge case, when delegate call is made directly by test contract.
+        // In such case, `top_cheated_data()` will have a default value (i.e. all fields are None),
```
Comment (cptartur) [inline crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/execution_utils.rs:20]: Maybe a reference to a function that is setting it in normal case would be helpful here.

File: docs/src/testing/testing-contract-internals.md [hunk 7]
```diff
@@ -82,6 +82,10 @@ We use the `SafeLibraryDispatcher` like this:
 ```rust
 {{#include ../../listings/testing_contract_internals/tests/using_library_calls.cairo}}
 ```
+
+> 📝 **Note**
+> Library calls don't affect [cheat spans](https://foundry-rs.github.io/starknet-foundry/testing/using-cheatcodes.html#setting-cheatcode-span) to progress. As a result,  the cheated data is still in effect when making subsequent library calls.
```
Comment (cptartur) [inline docs/src/testing/testing-contract-internals.md]:
```suggestion
> Library calls don't trigger [cheat spans](https://foundry-rs.github.io/starknet-foundry/testing/using-cheatcodes.html#setting-cheatcode-span) progression. As a result, the cheatcode is still in effect when making subsequent calls.
```

File: docs/src/testing/using-cheatcodes.md [hunk 8]
```diff
@@ -186,6 +186,9 @@ To better understand the functionality of `CheatSpan`, here's a full example:
 {{#include ../../listings/using_cheatcodes_others/tests/caller_address/span.cairo}}
 ```
 
+> 📝 **Note**
```
Comment (cptartur) [inline docs/src/testing/using-cheatcodes.md:189]: Same suggestion as above

**Review by cptartur — APPROVED** [review]
Left some comments, but otherwise good to merge

File: docs/src/testing/testing-contract-internals.md [hunk 9]
```diff
@@ -82,6 +82,10 @@ We use the `SafeLibraryDispatcher` like this:
 ```rust
 {{#include ../../listings/testing_contract_internals/tests/using_library_calls.cairo}}
 ```
+
+> 📝 **Note**
+> Library calls don't trigger [cheat spans](https://foundry-rs.github.io/starknet-foundry/testing/using-cheatcodes.html#setting-cheatcode-span) progression. As a result, the cheatcode is still in effect when making subsequent library calls.
+
```
Comment (MKowalski8) [inline docs/src/testing/testing-contract-internals.md:89]: What would you say to make it display like this:

Image

Now it is in one line:

Image

Also with this warning below it can look better with the `Note` seperation?

Of course, if you prefer the current approach, then leave it as it is.

File: docs/src/testing/using-cheatcodes.md [hunk 10]
```diff
@@ -186,6 +186,9 @@ To better understand the functionality of `CheatSpan`, here's a full example:
...
+> [Library calls](https://foundry-rs.github.io/starknet-foundry/testing/testing-contract-internals.html#using-library-calls-with-the-test-state-context) don't trigger cheat spans progression. As a result, the cheatcode is still in effect when making subsequent library calls.
```
Comment (MKowalski8) [inline docs/src/testing/using-cheatcodes.md:191]: Same as above

```suggestion
> 📝 **Note**
>
> [Library calls](https://foundry-rs.github.io/starknet-foundry/testing/testing-contract-internals.html#using-library-calls-with-the-test-state-context) don't trigger cheat spans progression. As a result, the cheatcode is still in effect when making subsequent library calls.
```

**Review by MKowalski8 — APPROVED** [review]
Please look to suggestion, but overall it looks good

File: docs/src/testing/testing-contract-internals.md (hunk 9 above)
Comment (franciszekjob) [reply docs/src/testing/testing-contract-internals.md:89]: Yes, there should be newline after Note, thanks.
